compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFoldDmaWaits.cpp

```diff
@@ -152,6 +153,147 @@ LogicalResult foldDmaWaits(const AMDAIE::AMDAIEDeviceModel &deviceModel,
   return success();
 }
 
+struct DmaColumnBatch {
+  uint32_t row;
+  uint32_t channel;
+  AMDAIE::DMAChannelDir direction;
+
+  // Sorted by column.
+  std::map<uint32_t, AMDAIE::NpuDmaWaitOp> colWaitOpMap;
+};
+
+/// Updates a batch of asynchronous DMA wait operations by combining their
+/// async tokens into a single NpuDmaWaitOp.
+void updateColumnBatchTokens(
+    IRRewriter &rewriter,
+    std::map<uint32_t, AMDAIE::NpuDmaWaitOp> &colWaitOpMap) {
+  if (colWaitOpMap.size() < 2) return;
+
+  // Check if there is any discontinuity in the columns, and if so, split into
+  // separate batches.
+  SmallVector<SmallVector<AMDAIE::NpuDmaWaitOp>> waitOpsList;
+  uint32_t prevCol = 0;
+  for (auto &entry : colWaitOpMap) {
+    uint32_t col = entry.first;
+    AMDAIE::NpuDmaWaitOp waitOp = entry.second;
+    if (waitOpsList.empty() || col != prevCol + 1) {
+      waitOpsList.push_back({});
+    }
+    waitOpsList.back().push_back(waitOp);
+    prevCol = col;
+  }
+
+  for (SmallVector<AMDAIE::NpuDmaWaitOp> &waitOps : waitOpsList) {
+    // For each batch, combine the async tokens into a single NpuDmaWaitOp.
+    SmallVector<Value> asyncTokens;
+    for (AMDAIE::NpuDmaWaitOp waitOp : waitOps) {
+      asyncTokens.append(waitOp.getAsyncTokens().begin(),
+                         waitOp.getAsyncTokens().end());
+    }
+    rewriter.setInsertionPointAfter(waitOps.back());
+    rewriter.create<AMDAIE::NpuDmaWaitOp>(waitOps.back().getLoc(), asyncTokens);
+    for (AMDAIE::NpuDmaWaitOp waitOp : waitOps) {
+      rewriter.eraseOp(waitOp);
+    }
+  }
+}
+
+/// Utility function to determine if a DMA wait operation can be folded.
+/// This is achieved by verifying whether it shares the same row, channel,
+/// and direction with preceding wait operations.
```

Comment:
You don't necessarily need to check whether they share the same row, channel, direction and you could just batch all tokens until some other operation touches the same connection and then leave the column batch discovery to `ControlCodeToTransactionLowering`.

Reply:
Similar reason as before, I feel maintaining a clear one-to-one correspondence between `dma_wait` and `TCT_sync` probably is easier for debugging.

Reply:
There is not such one-to-one correspondence assumed in the `NpuDmaWaitOp` as you can wait on multiple tokens at the same time, possibly resulting in multiple TCT syncs. This is supported in the current implementation and enables more compact IR as well as the optimization you're adding here. I don't see a need to limit  `NpuDmaWaitOp` to a single TCT sync.

Reply:
I understand it allows a more compact IR, but inside `ControlCodeToTransaction`, `NpuDmaWaitOp` is the only operation that does not have one-to-one correspondence. Unlike a DMA copy, which follows a clear transformation path (`dmaCpy` → `halfDmaCpy` → `pushToQueue` -> `TXN_OPC_WRITE`), I feel it is unclear what `NpuDmaWaitOp` represents, as its definition is a bit far away from `TCTSync`. 

This lack of one-to-one correspondence also puts the `FoldDmaWaits` pass in an awkward position. Although the pass reduces the number of `NpuDmaWaitOp`, there is no guarantee that this will result in fewer `TCTSync` operations.

May I suggest introducing an intermediate operation between `NpuDmaWaitOp` and `TCTSync`? This intermediate operation would only allow multiple tokens if they can be grouped into a single `TCTSync`. The `FoldDmaWaits` pass could then focus on optimizing at this intermediate level, while `ControlCodeToTransaction` only handles one-to-one conversion.

Reply:
> May I suggest introducing an intermediate operation between NpuDmaWaitOp and TCTSync? This intermediate operation would only allow multiple tokens if they can be grouped into a single TCTSync. The FoldDmaWaits pass could then focus on optimizing at this intermediate level, while ControlCodeToTransaction only handles one-to-one conversion.

I think a lower-level TCT-like operation makes sense, with `NpuDmaWaitOp` to `TCTSyncOp` lowering part of the ControlCodeLowering pass. If you go this route, create a separate PR though that introduces `TCTSyncOp` with rationale you have above. 

I am not convinced though that `FoldDmaWaits` should operate on the `TCTSyncOps` though. I kind of like that it's done at the higher level and just tries to batch as many syncs as possible because:
1. As a by-product, it results in more compact IR with clear boundaries
2. If new/future sync optimization opportunities become available (different channels), this pass can just stay the same and more low-level checks/transformations can take care of finding and leveraging the new capabilities within a batch. 
3. I don't see a case now or in the future in which we wouldn't want to batch waits.

> Although the pass reduces the number of NpuDmaWaitOp, there is no guarantee that this will result in fewer TCTSync operations.

I think that's fine, the DMA wait batching is general enough to potentially help in a lot of different (future) cases, without any obvious disadvantages, see reasoning above.

Reply:
Sure, I will make a separate PR for adding such `TCTSyncOp`.

Reply:
implemented in #990.